Unexpected escape analysis behavior for interpolated string with variable of local scope

The escape analysis behavior for the following interpolated string cases is unexpected:
```csharp
using System;
using System.Runtime.CompilerServices;

[InterpolatedStringHandler]
ref struct CustomHandler
{
    public CustomHandler(int literalLength, int formattedCount) { }
    public void AppendFormatted(Span<char> s) { }
}

class Program
{
    static CustomHandler F1()
    {
        Span<char> s = stackalloc char[10];
        CustomHandler h1 = $"{s}";
        return h1; // no error
    }

    static void F2()
    {
        Span<char> s = stackalloc char[10];
        F3($"{s}"); // error CS8352: ... may expose variables outside declaration scope
    }

    static void F3(ref CustomHandler handler) { }
}
```

For `F1()`, no error is reported returning `h1` although it may capture the local `Span<char>` from `$"{s}"`.

For `F2()`, an error is reported calling `F3($"{s}")`, although it seems no references should escape `F3()`.

--- a/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
@@ -1421,7 +1421,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             uint escapeScope = GetValEscape(data.Construction, scopeOfTheContainingExpression);
 
             var arguments = ArrayBuilder<BoundExpression>.GetInstance();
-            GetInterpolatedStringHandlerConversionArguments(expression, arguments);
+            GetInterpolatedStringHandlerArgumentsForEscape(expression, arguments);
 
             foreach (var argument in arguments)
             {
@@ -3823,7 +3823,7 @@ moreArguments:
             CheckValEscape(expression.Syntax, data.Construction, escapeFrom, escapeTo, checkingReceiver: false, diagnostics);
 
             var arguments = ArrayBuilder<BoundExpression>.GetInstance();
-            GetInterpolatedStringHandlerConversionArguments(expression, arguments);
+            GetInterpolatedStringHandlerArgumentsForEscape(expression, arguments);
 
             bool result = true;
             foreach (var argument in arguments)
@@ -3839,19 +3839,19 @@ moreArguments:
             return result;
         }
 
-        private static void GetInterpolatedStringHandlerConversionArguments(BoundExpression expression, ArrayBuilder<BoundExpression> arguments)
+        private void GetInterpolatedStringHandlerArgumentsForEscape(BoundExpression expression, ArrayBuilder<BoundExpression> arguments)
         {
             while (true)
             {
                 switch (expression)
                 {
                     case BoundBinaryOperator binary:
-                        getParts((BoundInterpolatedString)binary.Right, arguments);
+                        getParts((BoundInterpolatedString)binary.Right);
                         expression = binary.Left;
                         break;
 
                     case BoundInterpolatedString interpolatedString:
-                        getParts(interpolatedString, arguments);
+                        getParts(interpolatedString);
                         return;
 
                     default:
@@ -3859,7 +3859,7 @@ moreArguments:
                 }
             }
 
-            static void getParts(BoundInterpolatedString interpolatedString, ArrayBuilder<BoundExpression> arguments)
+            void getParts(BoundInterpolatedString interpolatedString)
             {
                 foreach (var part in interpolatedString.Parts)
                 {
@@ -3872,6 +3872,13 @@ moreArguments:
 
                     // The interpolation component is always the first argument to the method, and it was not passed by name
                     // so there can be no reordering.
+
+                    if (UseUpdatedEscapeRules &&
+                        call.Method.Parameters[0].EffectiveScope == DeclarationScope.ValueScoped)
+                    {
+                        continue;
+                    }
+
                     arguments.Add(call.Arguments[0]);
                 }
             }

--- a/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
@@ -1414,9 +1414,24 @@ namespace Microsoft.CodeAnalysis.CSharp
         }
 
         internal uint GetInterpolatedStringHandlerConversionEscapeScope(
-            InterpolatedStringHandlerData data,
+            BoundExpression expression,
             uint scopeOfTheContainingExpression)
-            => GetValEscape(data.Construction, scopeOfTheContainingExpression);
+        {
+            var data = expression.GetInterpolatedStringHandlerData();
+            uint escapeScope = GetValEscape(data.Construction, scopeOfTheContainingExpression);
+
+            var arguments = ArrayBuilder<BoundExpression>.GetInstance();
+            GetInterpolatedStringHandlerConversionArguments(expression, arguments);
+
+            foreach (var argument in arguments)
+            {
+                uint argEscape = GetValEscape(argument, scopeOfTheContainingExpression);
+                escapeScope = Math.Max(escapeScope, argEscape);
+            }
+
+            arguments.Free();
+            return escapeScope;
+        }
 
 #nullable enable
 
@@ -2995,8 +3010,7 @@ moreArguments:
 
                     if (conversion.ConversionKind == ConversionKind.InterpolatedStringHandler)
                     {
-                        var data = conversion.Operand.GetInterpolatedStringHandlerData();
-                        return GetInterpolatedStringHandlerConversionEscapeScope(data, scopeOfTheContainingExpression);
+                        return GetInterpolatedStringHandlerConversionEscapeScope(conversion.Operand, scopeOfTheContainingExpression);
                     }
 
                     return GetValEscape(conversion.Operand, scopeOfTheContainingExpression);
@@ -3808,33 +3822,44 @@ moreArguments:
 
             CheckValEscape(expression.Syntax, data.Construction, escapeFrom, escapeTo, checkingReceiver: false, diagnostics);
 
+            var arguments = ArrayBuilder<BoundExpression>.GetInstance();
+            GetInterpolatedStringHandlerConversionArguments(expression, arguments);
+
+            bool result = true;
+            foreach (var argument in arguments)
+            {
+                if (!CheckValEscape(argument.Syntax, argument, escapeFrom, escapeTo, checkingReceiver: false, diagnostics))
+                {
+                    result = false;
+                    break;
+                }
+            }
+
+            arguments.Free();
+            return result;
+        }
+
+        private static void GetInterpolatedStringHandlerConversionArguments(BoundExpression expression, ArrayBuilder<BoundExpression> arguments)
+        {
             while (true)
             {
                 switch (expression)
                 {
                     case BoundBinaryOperator binary:
-                        if (!checkParts((BoundInterpolatedString)binary.Right))
-                        {
-                            return false;
-                        }
-
+                        getParts((BoundInterpolatedString)binary.Right, arguments);
                         expression = binary.Left;
-                        continue;
+                        break;
 
                     case BoundInterpolatedString interpolatedString:
-                        if (!checkParts(interpolatedString))
-                        {
-                            return false;
-                        }
-
-                        return true;
+                        getParts(interpolatedString, arguments);
+                        return;
 
                     default:
                         throw ExceptionUtilities.UnexpectedValue(expression.Kind);
                 }
             }
 
-            bool checkParts(BoundInterpolatedString interpolatedString)
+            static void getParts(BoundInterpolatedString interpolatedString, ArrayBuilder<BoundExpression> arguments)
             {
                 foreach (var part in interpolatedString.Parts)
                 {
@@ -3847,16 +3872,8 @@ moreArguments:
 
                     // The interpolation component is always the first argument to the method, and it was not passed by name
                     // so there can be no reordering.
-                    var argument = call.Arguments[0];
-                    var success = CheckValEscape(argument.Syntax, argument, escapeFrom, escapeTo, checkingReceiver: false, diagnostics);
-
-                    if (!success)
-                    {
-                        return false;
-                    }
+                    arguments.Add(call.Arguments[0]);
                 }
-
-                return true;
             }
         }
 

--- a/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
+++ b/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs
@@ -1835,7 +1835,7 @@ moreArguments:
                 if (UseUpdatedEscapeRules)
                 {
                     // SPEC: For a given argument `a` that is passed to parameter `p`:
-                    // SPEC:  1. ...
+                    // SPEC: 1. ...
                     // SPEC: 2. If `p` is `scoped` then `a` does not contribute *safe-to-escape* when considering arguments.
                     if (parameter?.EffectiveScope == DeclarationScope.ValueScoped)
                     {
@@ -3873,6 +3873,9 @@ moreArguments:
                     // The interpolation component is always the first argument to the method, and it was not passed by name
                     // so there can be no reordering.
 
+                    // SPEC: For a given argument `a` that is passed to parameter `p`:
+                    // SPEC: 1. ...
+                    // SPEC: 2. If `p` is `scoped` then `a` does not contribute *safe-to-escape* when considering arguments.
                     if (UseUpdatedEscapeRules &&
                         call.Method.Parameters[0].EffectiveScope == DeclarationScope.ValueScoped)
                     {

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
@@ -13425,7 +13425,7 @@ AppendFormatted");
         [Theory]
         [InlineData(@"$""{s}""")]
         [InlineData(@"$""{s}"" + $""""")]
-        public void RefEscape_01(string expression)
+        public void RefEscape_01A(string expression)
         {
             var code = @"
 using System;
@@ -13454,6 +13454,35 @@ public ref struct CustomHandler
                 //         return $"{s}";
                 Diagnostic(ErrorCode.ERR_EscapeVariable, "s").WithArguments("s").WithLocation(17, 19)
             );
+        }
+
+        // As above but with scoped parameter in AppendFormatted().
+        [WorkItem(63262, "https://github.com/dotnet/roslyn/issues/63262")]
+        [Theory]
+        [InlineData(@"$""{s}""")]
+        [InlineData(@"$""{s}"" + $""""")]
+        public void RefEscape_01B(string expression)
+        {
+            var code = @"
+using System;
+using System.Runtime.CompilerServices;
+
+[InterpolatedStringHandler]
+public ref struct CustomHandler
+{
+    public CustomHandler(int literalLength, int formattedCount) { }
+
+    public void AppendFormatted(scoped Span<char> s) { }
+
+    public static CustomHandler M()
+    {
+        Span<char> s = stackalloc char[10];
+        return " + expression + @";
+    }
+}
+";
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics();
         }
 
         [Theory]
@@ -13957,7 +13986,7 @@ class Program
     {
         Span<char> s = stackalloc char[10];
         CustomHandler h1 = $""{s}"";
-        return h1; // 1
+        return h1;
     }
     static CustomHandler F2()
     {
@@ -13971,17 +14000,14 @@ class Program
         Span<char> s = stackalloc char[10];
         scoped CustomHandler h3 = new CustomHandler(0, 1);
         h3.AppendFormatted(s);
-        return h3; // 2
+        return h3; // 1
     }
 }
 ";
             var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
             comp.VerifyDiagnostics(
-                // (15,16): error CS8352: Cannot use variable 'h1' in this context because it may expose referenced variables outside of their declaration scope
-                //         return h1; // 1
-                Diagnostic(ErrorCode.ERR_EscapeVariable, "h1").WithArguments("h1").WithLocation(15, 16),
                 // (29,16): error CS8352: Cannot use variable 'h3' in this context because it may expose referenced variables outside of their declaration scope
-                //         return h3; // 2
+                //         return h3; // 1
                 Diagnostic(ErrorCode.ERR_EscapeVariable, "h3").WithArguments("h3").WithLocation(29, 16));
         }
 

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
@@ -13819,7 +13819,7 @@ class Program
         [Theory]
         [InlineData(LanguageVersion.CSharp10)]
         [InlineData(LanguageVersion.CSharp11)]
-        public void RefEscape_11(LanguageVersion languageVersion)
+        public void RefEscape_11A(LanguageVersion languageVersion)
         {
             var code =
 @"using System;
@@ -13832,15 +13832,63 @@ ref struct CustomHandler
 }
 class Program
 {
-    static void Main()
+    static void F1()
     {
         Span<char> s = stackalloc char[10];
         M($""{s}"");
+    }
+    static void F2()
+    {
+        Span<char> s = stackalloc char[10];
+        CustomHandler h2 = new CustomHandler(0, 1);
+        h2.AppendFormatted(s); // 1
+        M(ref h2);
     }
     static void M(ref CustomHandler handler) { }
 }
 ";
             var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion), targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics(
+                // (20,9): error CS8350: This combination of arguments to 'CustomHandler.AppendFormatted(Span<char>)' is disallowed because it may expose variables referenced by parameter 's' outside of their declaration scope
+                //         h2.AppendFormatted(s); // 1
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "h2.AppendFormatted(s)").WithArguments("CustomHandler.AppendFormatted(System.Span<char>)", "s").WithLocation(20, 9),
+                // (20,28): error CS8352: Cannot use variable 's' in this context because it may expose referenced variables outside of their declaration scope
+                //         h2.AppendFormatted(s); // 1
+                Diagnostic(ErrorCode.ERR_EscapeVariable, "s").WithArguments("s").WithLocation(20, 28));
+        }
+
+        // As above but with scoped parameter in AppendFormatted().
+        [WorkItem(63262, "https://github.com/dotnet/roslyn/issues/63262")]
+        [Fact]
+        public void RefEscape_11B()
+        {
+            var code =
+@"using System;
+using System.Runtime.CompilerServices;
+[InterpolatedStringHandler]
+ref struct CustomHandler
+{
+    public CustomHandler(int literalLength, int formattedCount) { }
+    public void AppendFormatted(scoped Span<char> s) { }
+}
+class Program
+{
+    static void F1()
+    {
+        Span<char> s = stackalloc char[10];
+        M($""{s}"");
+    }
+    static void F2()
+    {
+        Span<char> s = stackalloc char[10];
+        CustomHandler h2 = new CustomHandler(0, 1);
+        h2.AppendFormatted(s);
+        M(ref h2);
+    }
+    static void M(ref CustomHandler handler) { }
+}
+";
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
             comp.VerifyDiagnostics();
         }
 
@@ -13848,7 +13896,7 @@ class Program
         [Theory]
         [InlineData(LanguageVersion.CSharp10)]
         [InlineData(LanguageVersion.CSharp11)]
-        public void RefEscape_12(LanguageVersion languageVersion)
+        public void RefEscape_12A(LanguageVersion languageVersion)
         {
             var code =
 @"using System;
@@ -13865,7 +13913,51 @@ class Program
     {
         Span<char> s = stackalloc char[10];
         CustomHandler h1 = $""{s}"";
-        return h1;
+        return h1; // 1
+    }
+    static CustomHandler F2()
+    {
+        Span<char> s = stackalloc char[10];
+        CustomHandler h2 = new CustomHandler(0, 1);
+        h2.AppendFormatted(s); // 2
+        return h2;
+    }
+}
+";
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion), targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics(
+                // (15,16): error CS8352: Cannot use variable 'h1' in this context because it may expose referenced variables outside of their declaration scope
+                //         return h1; // 1
+                Diagnostic(ErrorCode.ERR_EscapeVariable, "h1").WithArguments("h1").WithLocation(15, 16),
+                // (21,9): error CS8350: This combination of arguments to 'CustomHandler.AppendFormatted(Span<char>)' is disallowed because it may expose variables referenced by parameter 's' outside of their declaration scope
+                //         h2.AppendFormatted(s); // 2
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "h2.AppendFormatted(s)").WithArguments("CustomHandler.AppendFormatted(System.Span<char>)", "s").WithLocation(21, 9),
+                // (21,28): error CS8352: Cannot use variable 's' in this context because it may expose referenced variables outside of their declaration scope
+                //         h2.AppendFormatted(s); // 2
+                Diagnostic(ErrorCode.ERR_EscapeVariable, "s").WithArguments("s").WithLocation(21, 28));
+        }
+
+        // As above but with scoped parameter in AppendFormatted().
+        [WorkItem(63262, "https://github.com/dotnet/roslyn/issues/63262")]
+        [Fact]
+        public void RefEscape_12B()
+        {
+            var code =
+@"using System;
+using System.Runtime.CompilerServices;
+[InterpolatedStringHandler]
+ref struct CustomHandler
+{
+    public CustomHandler(int literalLength, int formattedCount) { }
+    public void AppendFormatted(scoped Span<char> s) { }
+}
+class Program
+{
+    static CustomHandler F1()
+    {
+        Span<char> s = stackalloc char[10];
+        CustomHandler h1 = $""{s}"";
+        return h1; // 1
     }
     static CustomHandler F2()
     {
@@ -13874,19 +13966,23 @@ class Program
         h2.AppendFormatted(s);
         return h2;
     }
+    static CustomHandler F3()
+    {
+        Span<char> s = stackalloc char[10];
+        scoped CustomHandler h3 = new CustomHandler(0, 1);
+        h3.AppendFormatted(s);
+        return h3; // 2
+    }
 }
 ";
-            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion), targetFramework: TargetFramework.NetCoreApp);
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
             comp.VerifyDiagnostics(
                 // (15,16): error CS8352: Cannot use variable 'h1' in this context because it may expose referenced variables outside of their declaration scope
-                //         return h1;
+                //         return h1; // 1
                 Diagnostic(ErrorCode.ERR_EscapeVariable, "h1").WithArguments("h1").WithLocation(15, 16),
-                // (21,9): error CS8350: This combination of arguments to 'CustomHandler.AppendFormatted(Span<char>)' is disallowed because it may expose variables referenced by parameter 's' outside of their declaration scope
-                //         h2.AppendFormatted(s);
-                Diagnostic(ErrorCode.ERR_CallArgMixing, "h2.AppendFormatted(s)").WithArguments("CustomHandler.AppendFormatted(System.Span<char>)", "s").WithLocation(21, 9),
-                // (21,28): error CS8352: Cannot use variable 's' in this context because it may expose referenced variables outside of their declaration scope
-                //         h2.AppendFormatted(s);
-                Diagnostic(ErrorCode.ERR_EscapeVariable, "s").WithArguments("s").WithLocation(21, 28));
+                // (29,16): error CS8352: Cannot use variable 'h3' in this context because it may expose referenced variables outside of their declaration scope
+                //         return h3; // 2
+                Diagnostic(ErrorCode.ERR_EscapeVariable, "h3").WithArguments("h3").WithLocation(29, 16));
         }
 
         [Theory, WorkItem(54703, "https://github.com/dotnet/roslyn/issues/54703")]

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
@@ -13815,6 +13815,80 @@ class Program
                 Diagnostic(ErrorCode.ERR_EscapeCall, @"$""{1}""").WithArguments("CustomHandler.CustomHandler(int, int, ref S)", "s").WithLocation(17, 18));
         }
 
+        [WorkItem(63262, "https://github.com/dotnet/roslyn/issues/63262")]
+        [Theory]
+        [InlineData(LanguageVersion.CSharp10)]
+        [InlineData(LanguageVersion.CSharp11)]
+        public void RefEscape_11(LanguageVersion languageVersion)
+        {
+            var code =
+@"using System;
+using System.Runtime.CompilerServices;
+[InterpolatedStringHandler]
+ref struct CustomHandler
+{
+    public CustomHandler(int literalLength, int formattedCount) { }
+    public void AppendFormatted(Span<char> s) { }
+}
+class Program
+{
+    static void Main()
+    {
+        Span<char> s = stackalloc char[10];
+        M($""{s}"");
+    }
+    static void M(ref CustomHandler handler) { }
+}
+";
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion), targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics();
+        }
+
+        [WorkItem(63262, "https://github.com/dotnet/roslyn/issues/63262")]
+        [Theory]
+        [InlineData(LanguageVersion.CSharp10)]
+        [InlineData(LanguageVersion.CSharp11)]
+        public void RefEscape_12(LanguageVersion languageVersion)
+        {
+            var code =
+@"using System;
+using System.Runtime.CompilerServices;
+[InterpolatedStringHandler]
+ref struct CustomHandler
+{
+    public CustomHandler(int literalLength, int formattedCount) { }
+    public void AppendFormatted(Span<char> s) { }
+}
+class Program
+{
+    static CustomHandler F1()
+    {
+        Span<char> s = stackalloc char[10];
+        CustomHandler h1 = $""{s}"";
+        return h1;
+    }
+    static CustomHandler F2()
+    {
+        Span<char> s = stackalloc char[10];
+        CustomHandler h2 = new CustomHandler(0, 1);
+        h2.AppendFormatted(s);
+        return h2;
+    }
+}
+";
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion), targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics(
+                // (15,16): error CS8352: Cannot use variable 'h1' in this context because it may expose referenced variables outside of their declaration scope
+                //         return h1;
+                Diagnostic(ErrorCode.ERR_EscapeVariable, "h1").WithArguments("h1").WithLocation(15, 16),
+                // (21,9): error CS8350: This combination of arguments to 'CustomHandler.AppendFormatted(Span<char>)' is disallowed because it may expose variables referenced by parameter 's' outside of their declaration scope
+                //         h2.AppendFormatted(s);
+                Diagnostic(ErrorCode.ERR_CallArgMixing, "h2.AppendFormatted(s)").WithArguments("CustomHandler.AppendFormatted(System.Span<char>)", "s").WithLocation(21, 9),
+                // (21,28): error CS8352: Cannot use variable 's' in this context because it may expose referenced variables outside of their declaration scope
+                //         h2.AppendFormatted(s);
+                Diagnostic(ErrorCode.ERR_EscapeVariable, "s").WithArguments("s").WithLocation(21, 28));
+        }
+
         [Theory, WorkItem(54703, "https://github.com/dotnet/roslyn/issues/54703")]
         [InlineData(@"$""{{ {i} }}""")]
         [InlineData(@"$""{{ "" + $""{i}"" + $"" }}""")]

--- a/src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
+++ b/src/Compilers/CSharp/Test/Semantic/Semantics/InterpolationTests.cs
@@ -13448,12 +13448,18 @@ public ref struct CustomHandler
 }
 ";
 
-            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
-            comp.VerifyDiagnostics(
+            var expectedDiagnostics = new[]
+            {
                 // (17,19): error CS8352: Cannot use variable 's' in this context because it may expose referenced variables outside of their declaration scope
                 //         return $"{s}";
                 Diagnostic(ErrorCode.ERR_EscapeVariable, "s").WithArguments("s").WithLocation(17, 19)
-            );
+            };
+
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, parseOptions: TestOptions.Regular10, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics(expectedDiagnostics);
+
+            comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics(expectedDiagnostics);
         }
 
         // As above but with scoped parameter in AppendFormatted().
@@ -13511,12 +13517,18 @@ public ref struct CustomHandler
 }
 ";
 
-            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
-            comp.VerifyDiagnostics(
+            var expectedDiagnostics = new[]
+            {
                 // (17,9): error CS8150: By-value returns may only be used in methods that return by value
                 //         return $"{s}";
                 Diagnostic(ErrorCode.ERR_MustHaveRefReturn, "return").WithLocation(17, 9)
-            );
+            };
+
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, parseOptions: TestOptions.Regular10, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics(expectedDiagnostics);
+
+            comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics(expectedDiagnostics);
         }
 
         [Theory]
@@ -13545,12 +13557,18 @@ public ref struct CustomHandler
 }
 ";
 
-            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
-            comp.VerifyDiagnostics(
+            var expectedDiagnostics = new[]
+            {
                 // (17,20): error CS8156: An expression cannot be used in this context because it may not be passed or returned by reference
                 //         return ref $"{s}";
                 Diagnostic(ErrorCode.ERR_RefReturnLvalueExpected, expression).WithLocation(17, 20)
-            );
+            };
+
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, parseOptions: TestOptions.Regular10, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics(expectedDiagnostics);
+
+            comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute }, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics(expectedDiagnostics);
         }
 
         [Theory]
@@ -13646,7 +13664,10 @@ public ref struct S1
 }
 ";
 
-            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, targetFramework: TargetFramework.NetCoreApp);
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, parseOptions: TestOptions.Regular10, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics();
+
+            comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, targetFramework: TargetFramework.NetCoreApp);
             comp.VerifyDiagnostics();
         }
 
@@ -13677,7 +13698,10 @@ public ref struct CustomHandler
 }
 ";
 
-            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, targetFramework: TargetFramework.NetCoreApp);
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, parseOptions: TestOptions.Regular10, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics();
+
+            comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, targetFramework: TargetFramework.NetCoreApp);
             comp.VerifyDiagnostics();
         }
 
@@ -13708,12 +13732,17 @@ public ref struct CustomHandler
 }
 ";
 
-            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, targetFramework: TargetFramework.NetCoreApp);
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, parseOptions: TestOptions.Regular10, targetFramework: TargetFramework.NetCoreApp);
+            comp.VerifyDiagnostics();
+
+            comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, targetFramework: TargetFramework.NetCoreApp);
             comp.VerifyDiagnostics();
         }
 
-        [Fact]
-        public void RefEscape_08()
+        [Theory]
+        [InlineData(LanguageVersion.CSharp10)]
+        [InlineData(LanguageVersion.CSharp11)]
+        public void RefEscape_08(LanguageVersion languageVersion)
         {
             var code = @"
 using System;
@@ -13740,7 +13769,7 @@ public ref struct CustomHandler
 }
 ";
 
-            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, targetFramework: TargetFramework.NetCoreApp);
+            var comp = CreateCompilation(new[] { code, InterpolatedStringHandlerAttribute, InterpolatedStringHandlerArgumentAttribute }, parseOptions: TestOptions.Regular.WithLanguageVersion(languageVersion), targetFramework: TargetFramework.NetCoreApp);
             comp.VerifyDiagnostics(
                 // (16,16): error CS8352: Cannot use variable 'c' in this context because it may expose referenced variables outside of their declaration scope
                 //         return c;

